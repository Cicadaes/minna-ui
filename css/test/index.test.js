/** @jest-environment node */

'use strict';

const postcss = require('postcss'); // eslint-disable-line import/no-extraneous-dependencies
const postcssMixins = require('postcss-mixins'); // eslint-disable-line import/no-extraneous-dependencies
const drawTriangle = require('../src/mixins/draw-triangle.js');
const fluidFontSize = require('../src/mixins/fluid-font-size.js');
const makeFontFace = require('../src/mixins/make-font-face.js');

/**
 * Compile a CSS mixin;
 * @param {Function} mixinFn
 * @param {string} opts
 * @returns {object}
 */
function compileMixin(mixinFn, opts) {
  const plugin = [postcssMixins({ mixins: { mixinFn } })];
  const css = `
    div {
      @mixin mixinFn ${opts};
    }
  `;
  return postcss(plugin).process(css, { from: undefined });
}

describe('Global CSS', () => {
  it.skip('should do something', () => {
    expect.hasAssertions();
    // TODO
  });
});

describe('Draw triangle mixin', () => {
  it('outputs top triangle', async () => {
    const output = drawTriangle(null, 'top', '1rem', '#000');
    expect(output).toHaveProperty('border-bottom');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(drawTriangle, 'top, 1rem, #000');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('outputs right triangle', async () => {
    const output = drawTriangle(null, 'right', '1rem', '#000');
    expect(output).toHaveProperty('border-left');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(drawTriangle, 'right, 1rem, #000');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('outputs bottom triangle', async () => {
    const output = drawTriangle(null, 'bottom', '1rem', '#000');
    expect(output).toHaveProperty('border-top');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(drawTriangle, 'bottom, 1rem, #000');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('outputs left triangle', async () => {
    const output = drawTriangle(null, 'left', '1rem', '#000');
    expect(output).toHaveProperty('border-right');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(drawTriangle, 'left, 1rem, #000');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('outputs triangle in pseudo element', async () => {
    const output = drawTriangle(null, 'top', '1rem', '#000', true);
    expect(output).toHaveProperty('&::before.border-bottom');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(drawTriangle, 'top, 1rem, #000, true');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('throws error when invalid direction', () => {
    expect(() => {
      drawTriangle(null, 'under', '1rem', '#000');
    }).toThrow();
  });
});

describe('Fluid font size mixin', () => {
  it('generates CSS rules', async () => {
    const output = fluidFontSize(null, '16px', '32px', '30em', '100em');
    expect(output).not.toHaveProperty('&.font-size');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(fluidFontSize, '16px, 32px, 30em, 100em');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('generates CSS rules in px', async () => {
    const output = fluidFontSize(null, '1rem', '2rem', '30em', '100em', true);
    expect(output).not.toHaveProperty('&.font-size');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(fluidFontSize, '1rem, 2rem, 30em, 100em, true');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('includes base font-size when minFontSize isn\'t browser default', async () => {
    const output = fluidFontSize(null, '1.5rem', '2rem', '30em', '100em');
    expect(output).toHaveProperty('&.font-size');

    const result = await compileMixin(fluidFontSize, '1.5rem, 2rem, 30em, 100em');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('includes base font-size when minFontSize isn\'t browser default in px', async () => {
    const output = fluidFontSize(null, '20px', '2rem', '30em', '100em');
    expect(output).toHaveProperty('&.font-size');

    const result = await compileMixin(fluidFontSize, '20px, 2rem, 30em, 100em');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('throws error when invalid size unit', () => {
    expect(() => {
      fluidFontSize(null, '16', '32pops', '30em', '100em');
    }).toThrow('Only px, rem, and em units are supported');
  });
});

describe('Make font face mixin', () => {
  it('generates CSS rules', async () => {
    const output = makeFontFace(null, 'Target', 'Target-Regular');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(makeFontFace, 'Target, Target-Regular');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });

  it('generates CSS rules with custom options', async () => {
    const output = makeFontFace(null, 'Target', 'Target-Bold', 'normal', 'bold', '/static/fonts');
    expect(output).toMatchSnapshot();

    const result = await compileMixin(makeFontFace, 'Target, Target-Bold, normal, bold, /static/fonts');
    expect(result.warnings).toHaveLength(0);
    expect(result.css).toMatchSnapshot();
  });
});
