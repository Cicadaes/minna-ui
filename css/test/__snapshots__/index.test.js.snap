// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draw triangle mixin outputs bottom triangle 1`] = `
Object {
  "border-left": "1rem solid transparent",
  "border-right": "1rem solid transparent",
  "border-top": "1rem solid #000",
  "height": 0,
  "width": 0,
}
`;

exports[`Draw triangle mixin outputs bottom triangle 2`] = `
"
    div {
      width: 0;
      height: 0;
      border-top: 1rem solid #000;
      border-right: 1rem solid transparent;
      border-left: 1rem solid transparent;
    }
  "
`;

exports[`Draw triangle mixin outputs left triangle 1`] = `
Object {
  "border-bottom": "1rem solid transparent",
  "border-right": "1rem solid #000",
  "border-top": "1rem solid transparent",
  "height": 0,
  "width": 0,
}
`;

exports[`Draw triangle mixin outputs left triangle 2`] = `
"
    div {
      width: 0;
      height: 0;
      border-top: 1rem solid transparent;
      border-right: 1rem solid #000;
      border-bottom: 1rem solid transparent;
    }
  "
`;

exports[`Draw triangle mixin outputs right triangle 1`] = `
Object {
  "border-bottom": "1rem solid transparent",
  "border-left": "1rem solid #000",
  "border-top": "1rem solid transparent",
  "height": 0,
  "width": 0,
}
`;

exports[`Draw triangle mixin outputs right triangle 2`] = `
"
    div {
      width: 0;
      height: 0;
      border-top: 1rem solid transparent;
      border-bottom: 1rem solid transparent;
      border-left: 1rem solid #000;
    }
  "
`;

exports[`Draw triangle mixin outputs top triangle 1`] = `
Object {
  "border-bottom": "1rem solid #000",
  "border-left": "1rem solid transparent",
  "border-right": "1rem solid transparent",
  "height": 0,
  "width": 0,
}
`;

exports[`Draw triangle mixin outputs top triangle 2`] = `
"
    div {
      width: 0;
      height: 0;
      border-right: 1rem solid transparent;
      border-bottom: 1rem solid #000;
      border-left: 1rem solid transparent;
    }
  "
`;

exports[`Draw triangle mixin outputs triangle in pseudo element 1`] = `
Object {
  "&::before": Object {
    "border-bottom": "1rem solid #000",
    "border-left": "1rem solid transparent",
    "border-right": "1rem solid transparent",
    "content": "\\"\\"",
    "display": "block",
    "height": 0,
    "left": "calc(50% - 1rem)",
    "position": "absolute",
    "top": "calc(-1 * 1rem)",
    "width": 0,
  },
}
`;

exports[`Draw triangle mixin outputs triangle in pseudo element 2`] = `
"
    div {
      &::before {
            position: absolute;
            content: \\"\\";
            display: block;
            left: calc(50% - 1rem);
            top: calc(-1 * 1rem);
            width: 0;
            height: 0;
            border-right: 1rem solid transparent;
            border-bottom: 1rem solid #000;
            border-left: 1rem solid transparent
      }
    }
  "
`;

exports[`Fluid font size mixin generates CSS rules 1`] = `
Object {
  "&": Object {
    "@media screen and (min-width: 100em)": Object {
      "font-size": "2em",
    },
    "@media screen and (min-width: 30em) and (max-width: 100em)": Object {
      "font-size": "calc(1em + 1 * (100vw - 30em) / 70);",
    },
  },
}
`;

exports[`Fluid font size mixin generates CSS rules 2`] = `
"
    div {
      & {
            @media screen and (min-width: 30em) and (max-width: 100em) {
                  font-size: calc(1em + 1 * (100vw - 30em) / 70);
            }
            @media screen and (min-width: 100em) {
                  font-size: 2em
            }
      }
    }
  "
`;

exports[`Fluid font size mixin generates CSS rules in px 1`] = `
Object {
  "&": Object {
    "@media screen and (min-width: 100em)": Object {
      "font-size": "32px",
    },
    "@media screen and (min-width: 30em) and (max-width: 100em)": Object {
      "font-size": "calc(16px + 16 * (100vw - 480px) / 1120);",
    },
  },
}
`;

exports[`Fluid font size mixin generates CSS rules in px 2`] = `
"
    div {
      & {
            @media screen and (min-width: 30em) and (max-width: 100em) {
                  font-size: calc(16px + 16 * (100vw - 480px) / 1120);
            }
            @media screen and (min-width: 100em) {
                  font-size: 32px
            }
      }
    }
  "
`;

exports[`Fluid font size mixin includes base font-size when minFontSize isn't browser default 1`] = `
"
    div {
      & {
            font-size: 1.5em;
            @media screen and (min-width: 30em) and (max-width: 100em) {
                  font-size: calc(1.5em + 0.5 * (100vw - 30em) / 70);
            }
            @media screen and (min-width: 100em) {
                  font-size: 2em
            }
      }
    }
  "
`;

exports[`Fluid font size mixin includes base font-size when minFontSize isn't browser default in px 1`] = `
"
    div {
      & {
            font-size: 1.25em;
            @media screen and (min-width: 30em) and (max-width: 100em) {
                  font-size: calc(1.25em + 0.75 * (100vw - 30em) / 70);
            }
            @media screen and (min-width: 100em) {
                  font-size: 2em
            }
      }
    }
  "
`;

exports[`Make font face mixin generates CSS rules 1`] = `
Object {
  "@font-face": Object {
    "font-display": "optional",
    "font-family": "Target",
    "font-style": "normal",
    "font-weight": "normal",
    "src": "local(\\"Target-Regular\\"),
      url(\\"/fonts/Target-Regular.woff2\\") format(\\"woff2\\"),
      url(\\"/fonts/Target-Regular.woff\\") format(\\"woff\\")",
  },
}
`;

exports[`Make font face mixin generates CSS rules 2`] = `
"
    div {
      @font-face {
            font-family: Target;
            src: local(\\"Target-Regular\\"),
      url(\\"/fonts/Target-Regular.woff2\\") format(\\"woff2\\"),
      url(\\"/fonts/Target-Regular.woff\\") format(\\"woff\\");
            font-style: normal;
            font-weight: normal;
            font-display: optional
      }
    }
  "
`;

exports[`Make font face mixin generates CSS rules with custom options 1`] = `
Object {
  "@font-face": Object {
    "font-display": "optional",
    "font-family": "Target",
    "font-style": "normal",
    "font-weight": "bold",
    "src": "local(\\"Target-Bold\\"),
      url(\\"/static/fonts/Target-Bold.woff2\\") format(\\"woff2\\"),
      url(\\"/static/fonts/Target-Bold.woff\\") format(\\"woff\\")",
  },
}
`;

exports[`Make font face mixin generates CSS rules with custom options 2`] = `
"
    div {
      @font-face {
            font-family: Target;
            src: local(\\"Target-Bold\\"),
      url(\\"/static/fonts/Target-Bold.woff2\\") format(\\"woff2\\"),
      url(\\"/static/fonts/Target-Bold.woff\\") format(\\"woff\\");
            font-style: normal;
            font-weight: bold;
            font-display: optional
      }
    }
  "
`;
